src/org/traccar/database/PermissionsManager.java

```diff
@@ -168,30 +225,52 @@ public void checkUserEnabled(long userId) throws SecurityException {
 
     public void checkUserUpdate(long userId, User before, User after) throws SecurityException {
         if (before.getAdmin() != after.getAdmin()
-                || before.getReadonly() != after.getReadonly()
-                || before.getDisabled() != after.getDisabled()
                 || before.getDeviceLimit() != after.getDeviceLimit()
+                || before.getUserLimit() != after.getUserLimit()) {
+            checkAdmin(userId);
+        }
+        if (before.getReadonly() != after.getReadonly()
+                || before.getDisabled() != after.getDisabled()
                 || !Objects.equals(before.getExpirationTime(), after.getExpirationTime())
                 || !Objects.equals(before.getToken(), after.getToken())) {
-            checkAdmin(userId);
+            if (userId == after.getId()) {
+                checkAdmin(userId);
+            }
+            if (!isAdmin(userId)) {
+                checkManager(userId);
+            }
         }
     }
 
     public void checkUser(long userId, long otherUserId) throws SecurityException {
-        if (userId != otherUserId) {
-            checkAdmin(userId);
+        if (userId != otherUserId && !isAdmin(userId)) {
+            checkManager(userId, otherUserId);
         }
     }
 
     public void checkGroup(long userId, long groupId) throws SecurityException {
-        if (!getGroupPermissions(userId).contains(groupId)) {
-            throw new SecurityException("Group access denied");
+        if (!getGroupPermissions(userId).contains(groupId) && !isAdmin(userId)) {
+            Iterator<Long> iterator = getUserPermissions(userId).iterator();
```

Comment:
I have not noticed, and theoretically there should not be any.
- first `if` in function is repeat previous behavior
- `getUserPermissions(userId)` will return empty collection for ordinary user.
- If i'm not mistaken `contains` works with hashed data.

The only longer than usual check will be for manager, but everything is cached and uses `HashSet`s, do not see how it can be optimized.

Only one thought is add 'isManager(userId)` before cycle.